File: ablation three.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import font_manager
import os
import seaborn as sns

# --- Dataset files ---
files = [
    r'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\ml-1m\grit_hypertune_results.xlsx',
    r'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\ml-100k\grit_hypertune_results.xlsx',
    r'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\i_a_s\grit_hypertune_results.xlsx',
    r'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\v_g\grit_hypertune_results.xlsx',
    r'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\c_a_v\grit_hypertune_results.xlsx'
]

# Map short names to display names
dataset_map = {
    'ml_1m': 'MovieLens 1M',
    'ml_100k': 'MovieLens 100K',
    'v_g': 'Video Games',
    'i_a_s': 'Industrial & Scientific',
    'c_a_v': 'CDs & Vinyl'
}
dataset_keys = ['ml_1m','ml_100k','v_g','i_a_s','c_a_v']
dataset_names = [dataset_map[k] for k in dataset_keys]
all_groups = set()
all_betas = set()
tables = []

for f in files:
    T = pd.read_excel(f, sheet_name='val')
    T['avgScore'] = (T['Recall@10'] + T['MRR@10']) / 2
    all_groups.update(T['num_groups'].unique())
    all_betas.update(T['beta'].unique())
    tables.append(T)

all_groups = sorted(list(all_groups))
all_betas = sorted(list(all_betas))

# --- Build matrices with max scores ---
group_mat = np.full((len(files), len(all_groups)), np.nan)
beta_mat  = np.full((len(files), len(all_betas)), np.nan)

for i, T in enumerate(tables):
    for j, g in enumerate(all_groups):
        mask = T['num_groups'] == g
        if mask.any():
            # Take the max score instead of mean
            group_mat[i, j] = T.loc[mask, 'avgScore'].max()
    for j, b in enumerate(all_betas):
        mask = T['beta'] == b
        if mask.any():
            beta_mat[i, j] = T.loc[mask, 'avgScore'].max()

# --- Color palette (normalized RGB) ---
palette = [
    "#FFF44F",  # Lemon Yellow
    "#FF6F91",  # Blush Pink
    "#87CEEB",  # Sky Blue
    "#8EEA8E",  # Light Green (0.56, 0.93, 0.56)
    "#FACCA0",  # Peach       (0.98, 0.80, 0.69)
    "#AAD5E6",  # Light Cyan  (0.67, 0.84, 0.90)
]

# --- Function to plot grouped bars without gaps ---
def plot_grouped_bars(mat, dataset_names, param_values, xlabel, title, palette, save_path=None):
    n_datasets, n_params = mat.shape

    # Convert matrix to long-format DataFrame for seaborn
    data = []
    for i in range(n_datasets):
        for j in range(n_params):
            data.append({
                "dataset": dataset_names[i],
                "param": str(param_values[j]),
                "value": mat[i, j]
            })
    df_long = pd.DataFrame(data)

    sns.set_theme(style="whitegrid")
    plt.figure(figsize=(8, 6))

    ax = sns.barplot(
        data=df_long,
        x="dataset",
        y="value",
        hue="param",
        palette=palette,
        edgecolor="black",
        linewidth=0.5,
        dodge=True,
        ci=None,
        width=0.5
    )

    # X-axis formatting
    ax.set_xticklabels(
        [lbl.get_text() for lbl in ax.get_xticklabels()],
        rotation=20, ha="center",
        fontsize=14, fontweight="bold", family="Times New Roman"
    )
    plt.xlabel(xlabel, fontsize=14, fontweight="bold", family="Times New Roman")
    plt.ylabel("Score", fontsize=14, fontweight="bold", family="Times New Roman")

    # Y-axis scaling: round up to nearest 0.02
    max_val = np.nanmax(mat)
    y_max = np.ceil((max_val + 0.02) / 0.02) * 0.02
    plt.yticks(
        np.arange(0, y_max + 0.001, 0.02),
        fontsize=14, fontweight="bold", family="Times New Roman"
    )
    plt.ylim(0, y_max)

    # Grid
    plt.grid(axis="y", linestyle="--", alpha=0.5)

    # Legend above
    plt.legend(
        loc="upper center", bbox_to_anchor=(0.5, 1.1), ncol=min(8, n_params),
        frameon=False,
        prop=font_manager.FontProperties(family="Times New Roman", size=14, weight="bold")
    )

    # Title (optional)
    # plt.title(title, fontsize=14, fontweight="bold", family="Times New Roman")

    plt.tight_layout()

    if save_path is not None:
        os.makedirs(os.path.dirname(save_path), exist_ok=True)
        plt.savefig(save_path, dpi=300, bbox_inches="tight")
        plt.close()
    else:
        plt.show()

# --- Plot ---
plot_grouped_bars(group_mat, dataset_names, all_groups, 'Dataset', 'Effect of κ across datasets', palette, save_path='experiments/ablation/combined/hyper_k.png')
plot_grouped_bars(beta_mat, dataset_names, all_betas, 'Dataset', 'Effect of β across datasets', palette, save_path='experiments/ablation/combined/hyper_beta.png')

File: ablation_three.m
%% Mesh Plot

filename = 'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\i_a_s\grit_hypertune_results.xlsx';
[Result, ~, raw] = xlsread(filename, 'val');

% Read table for easier column handling
T = readtable(filename, 'Sheet', 'val');

% Compute average score of Recall@10 and MRR@10
T.avgScore = (T.Recall_10 + T.MRR_10) / 2;

% Get unique dropout and attn_dropout values
dropout_vals = unique(T.dropout);
attn_vals = unique(T.attn_dropout);

% Initialize matrix for Z values
resultFinal = nan(length(attn_vals), length(dropout_vals));

% Fill matrix with average scores
for i = 1:length(attn_vals)
    for j = 1:length(dropout_vals)
        subset = T(T.attn_dropout == attn_vals(i) & T.dropout == dropout_vals(j), :);
        if ~isempty(subset)
            resultFinal(i, j) = mean(subset.avgScore);
        end
    end
end

% Create meshgrid for plotting
[X, Y] = meshgrid(1:length(dropout_vals), 1:length(attn_vals));

% Plot the surface
figure;
surf(X, Y, resultFinal);

% Set ticks/labels
set(gca, 'XTick', 1:length(dropout_vals), 'XTickLabel', string(dropout_vals));
set(gca, 'YTick', 1:length(attn_vals), 'YTickLabel', string(attn_vals));

% Font settings for axes
set(gca, 'FontName', 'Times New Roman', 'FontSize', 20, 'FontWeight', 'bold');
colormap('pink');
% Axis labels and title
xlabel('Dropout', 'FontName', 'Times New Roman', 'FontWeight', 'bold', 'FontSize', 20);
ylabel('Attention Dropout', 'FontName', 'Times New Roman', 'FontWeight', 'bold', 'FontSize', 20);
zlabel('Score', 'FontName', 'Times New Roman', 'FontWeight', 'bold', 'FontSize', 20);
% title('Mesh Plot of Dropout vs Attention Dropout', 'FontName', 'Times New Roman', 'FontWeight', 'bold', 'FontSize', 15);
box on;


% %% Contour Plot
% 
% filename = 'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\c_a_v\grit_hypertune_results.xlsx';
% 
% % Read table
% T = readtable(filename, 'Sheet', 'val');
% 
% % Compute average score of Recall@10 and MRR@10
% T.avgScore = (T.Recall_10 + T.MRR_10) / 2;
% 
% % Get unique num_groups and beta values
% groups_vals = unique(T.num_groups);
% beta_vals   = unique(T.beta);
% 
% % Initialize result matrix
% resultFinal = nan(length(groups_vals), length(beta_vals));
% 
% % Fill matrix with mean values
% for i = 1:length(groups_vals)
%     for j = 1:length(beta_vals)
%         subset = T(T.num_groups == groups_vals(i) & T.beta == beta_vals(j), :);
%         if ~isempty(subset)
%             resultFinal(i, j) = mean(subset.avgScore);
%         end
%     end
% end
% 
% % Create meshgrid
% [X, Y] = meshgrid(beta_vals, groups_vals);
% 
% % Use indices for equal spacing, but label with actual values
% contourf(1:length(beta_vals), 1:length(groups_vals), resultFinal, 20, 'LineColor', 'w');
% hcb = colorbar;
% 
% % Set ticks to actual values
% set(gca, 'XTick', 1:length(beta_vals), 'XTickLabel', string(beta_vals));
% set(gca, 'YTick', 1:length(groups_vals), 'YTickLabel', string(groups_vals));
% 
% % Colormap (inverse bone)
% colormap(bone);
% 
% % Font settings
% set(gca, 'FontName', 'Times New Roman', 'FontSize', 15);
% set(hcb, 'FontName', 'Times New Roman', 'FontSize', 15, 'FontWeight', 'bold');
% 
% % Axis labels
% xlabel('\beta', 'FontName', 'Times New Roman', 'FontWeight', 'bold', 'FontSize', 15);
% ylabel('\kappa', 'FontName', 'Times New Roman', 'FontWeight', 'bold', 'FontSize', 15);
% 
% % Optional title
% % title('Contour Plot of Avg(Recall@10, MRR@10)', ...
% %     'FontName', 'Times New Roman', 'FontWeight', 'bold', 'FontSize', 15);
% 
% %% Bar Plot
% % List of dataset files
% files = {
%     'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\c_a_v\grit_hypertune_results.xlsx';
%     'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\i_a_s\grit_hypertune_results.xlsx';
%     'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\ml-1m\grit_hypertune_results.xlsx';
%     'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\ml-100k\grit_hypertune_results.xlsx';
%     'D:\Working\9. GrIT\Codes\experiments\hyperparameter_tuning\v_g\grit_hypertune_results.xlsx';
% };
% 
% datasets = cellfun(@(f) erase(extractAfter(f, 'results_'), '.xlsx'), files, 'UniformOutput', false);
% 
% % Initialize storage
% allGroups = [];
% allBetas  = [];
% 
% % First, collect all unique groups and betas across datasets
% for f = 1:length(files)
%     T = readtable(files{f}, 'Sheet', 'val');
%     T.avgScore = (T.Recall_10 + T.MRR_10) / 2;
%     allGroups = union(allGroups, unique(T.num_groups));
%     allBetas  = union(allBetas, unique(T.beta));
% end
% 
% % Matrices: rows = datasets, cols = unique groups/betas
% groupMat = nan(length(datasets), length(allGroups));
% betaMat  = nan(length(datasets), length(allBetas));
% 
% % Fill matrices
% for f = 1:length(files)
%     T = readtable(files{f}, 'Sheet', 'val');
%     T.avgScore = (T.Recall_10 + T.MRR_10) / 2;
% 
%     % groups
%     for j = 1:length(allGroups)
%         idx = T.num_groups == allGroups(j);
%         if any(idx)
%             groupMat(f, j) = mean(T.avgScore(idx));
%         end
%     end
% 
%     % betas
%     for j = 1:length(allBetas)
%         idx = T.beta == allBetas(j);
%         if any(idx)
%             betaMat(f, j) = mean(T.avgScore(idx));
%         end
%     end
% end
% % Colors (from your Python palette)
% palette = [
%     1.0, 1.0, 0.31;   % Lemon Yellow "#FFF44F"
%     1.0, 0.44, 0.57;  % Blush Pink "#FF6F91"
%     0.53, 0.81, 0.92; % Sky Blue "#87CEEB"
%     0.56, 0.93, 0.56; % Light Green "#8EEA8E"
%     1.0, 0.64, 0.0;   % Orange "#FFA300"
%     0.80, 0.60, 0.78; % Lavender "#CC99C9"
%     0.98, 0.80, 0.69; % Peach "#FACCA0"
%     0.67, 0.84, 0.90; % Light Cyan "#AAD5E6"
% ];
% 
% % Dataset display names
% datasetMap = containers.Map( ...
%     {'ml_1m','ml_100k','v_g','i_a_s','c_a_v'}, ...
%     {'MovieLens 1M','MovieLens 100K','Video Games','Industrial & Scientific','CDs & Vinyl'} ...
% );
% datasetKeys = {'ml_1m','ml_100k','v_g','i_a_s','c_a_v'};
% datasetNames = cellfun(@(k) datasetMap(k), datasetKeys, 'UniformOutput', false);
% 
% % Function to plot grouped bars
% function plotGroupedBars(mat, datasetNames, paramValues, xlabelText, titleText, palette)
%     figure;
%     b = bar(mat, 'grouped');  % grouped bars
% 
%     % Set colors
%     for j = 1:length(b)
%         b(j).FaceColor = palette(j,:);
%         b(j).EdgeColor = 'k';
%         b(j).LineWidth = 0.5;
%     end
% 
%     % Axes formatting
%     set(gca, 'XTick', 1:length(datasetNames), 'XTickLabel', datasetNames, ...
%         'FontName', 'Times New Roman', 'FontSize', 12);
%     xlabel(xlabelText, 'FontName', 'Times New Roman', 'FontSize', 15, 'FontWeight','bold');
%     ylabel('Score', 'FontName', 'Times New Roman', 'FontSize', 15, 'FontWeight','bold');
% 
%     % Grid
%     grid on;
%     ax = gca;
%     ax.YGrid = 'on';
%     ax.XGrid = 'off';
%     ax.GridLineStyle = '--';
%     ax.GridAlpha = 0.5;
% 
%     % Legend above plot
%     legend(string(paramValues), 'Location','northoutside','Orientation','horizontal', ...
%         'FontName','Times New Roman', 'FontSize', 10, 'Box','off');
% 
%     % title(titleText, 'FontName','Times New Roman', 'FontSize',12,'FontWeight','bold');
% end
% 
% % Plot groups (κ)
% plotGroupedBars(groupMat, datasetNames, allGroups, 'Dataset', 'Effect of \kappa across datasets', palette);
% 
% % Plot betas (β)
% plotGroupedBars(betaMat, datasetNames, allBetas, 'Dataset', 'Effect of \beta across datasets', palette);
